docs/DafnyRef/integration- rust/IntegrationRust.md

```diff
@@ -71,42 +71,114 @@ For maximum flexibility, Dafny follows these rules to encode types in Rust:
-| `C`, `C?` (for class, iterator C) | `*mut C`                |
-| (trait) `T`                   | (trait) `*mut dyn T`        |
-| `array<T>`                    | `*mut [T]`                  |
-| `array2<T>`                   | `*mut array2<T>`            |
+| `C`, `C?` (for class, iterator C) | `Ptr<C>`            (*) |
+| (trait) `T`                   | (trait) `Ptr<dyn T>`    (*) |
+| `array<T>`                    | `Ptr<[T]>`              (*) |
+| `array2<T>`                   | `Ptr<array2<T>>`    (*) (*) |
+| `c.x` if `c: C`               | `read!(c).x`            (*) |
+| `c.x := 2` if `c: C`          | `modify!(c).x = 2`      (*) |
+| `a[0]` if `a: array<T>`       | `read!(a)[0]`           (*) |
+| `a[1] := 2` if  `a: array<T>` | `modify!(a)[1] = 2`     (*) |
 
 (*) Defined in the Dafny runtime
 
+# Externs
+
+You can provide additional `*.rs` files to `dafny translate`, `dafny build` and even `dafny run` (via the `--input` option)
+All the elements imported from the Rust files are going to be available from the generated code via the module `_dafny_externs` that imports everything.
+
+The best way to see what you have to implement as an extern is to compile your code without it. You'll see that it will be necessary for the module `_dafny_extern` to use open a certain module, which you'd put in a file `external.rs` and pass it for compilation.
+
+When using simple 1-argument externs, the compilation follows Dafny conventions:
+* Static methods are static instance methods of a `pub struct _default {}` which is defined in the module itself, or must be defined externally if all static methods are labelled as externs.
+* Methods or functions with an extern attribute with 2 arguments are interpreted as follow: The first argument is a module name, and all the "." are rewritten to "::" to follow Rust's syntax. The second argument is the name of a static `pub fn` of that module.
+
+Let's assume you provide an additional input file `external.rs` with the following content:
+```
+pub mod rust {
+  pub mod module {
+    fn outsideMethod()
+  }
+}
+pub mod Test {
+  pub struct _default {}
+  impl _default {
+    pub fn extern_y() {
+    }
+  }
+}
+```
+Assuming that your Dafny program looks like:
+
+```
+
+module Test {
+  method {:extern "rust.module", "externalMethod"} outsideMethod()
+  method {:extern "extern_y"} dafny_y();
+}
+method Main() {
+  Test.outsideMethod();
+  Test.dafny_y();
+}
+```
+
+Dafny will generate the following for you:
+
+```
+pub mod external;         // from the additional external.rs input file
+
+pub mod _dafny_externs {  // from the additional external.rs input file
+  import opened external::*;
+}
+pub mod Test {
+  pub use super::_dafny_externs::Test::*; // Imports the _default
+}
+fn main() {
+  crate::rust::module::externalMethod();
+  super::Test::_default::extern_y();
+}
+```
+
+Note: 
+Note that if the extern module name contains a space, any dot after the space is not replaced by `::`.
```

Comment:
Why support that?

Reply:
In case someone wants to write a comment in an extern, and there is a dot in the comment, like:
```
{:extern "other.module/*We can do better. This is temporary*/", "functionName":}
```
We don't replace the second dot by `::`.
And if you think it makes it possible to make Rust injection, this makes it possible, like any other backend for now.

Reply:
I really don't want to encourage that. It implies that the semantics of externs must be to copy the string verbatim with no processing ever possible. Those kinds of comments can easily live elsewhere in a system.

Reply:
> It implies that the semantics of externs must be to copy the string verbatim with no processing ever possible
I removed this comment and moved it to Dafny-compiler-rust.dfy